### PR TITLE
Fix InListExpr  to return the correct number of rows

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/arrow-datafusion/issues/8600

## Rationale for this change

This was causing a panic for us when reading parquet files when pushing down filters . More details on https://github.com/apache/arrow-datafusion/issues/8600

## What changes are included in this PR?
1. Return the correct number of rows when evaluating InListExpr with constants

## Are these changes tested?
Yes

## Are there any user-facing changes?
Bug fix

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 1]
```diff
@@ -349,9 +349,10 @@ impl PhysicalExpr for InListExpr {
     }
 
     fn evaluate(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
+        let num_rows = batch.num_rows();
         let value = self.expr.evaluate(batch)?;
         let r = match &self.static_filter {
-            Some(f) => f.contains(value.into_array(1)?.as_ref(), self.negated)?,
+            Some(f) => f.contains(value.into_array(num_rows)?.as_ref(), self.negated)?,
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/in_list.rs:355]: This is the actual fix -- the rest of the PR is test

**Review by viirya — APPROVED** [review]

File: datafusion/physical-expr/src/expressions/in_list.rs [hunk 2]
```diff
@@ -349,9 +349,10 @@ impl PhysicalExpr for InListExpr {
...
             None => {
                 let value = value.into_array(batch.num_rows())?;
```
Comment (Dandandan) [inline datafusion/physical-expr/src/expressions/in_list.rs:357]: Maybe we can use the `let value = value.into_array(batch.num_rows())?;` for both arms here

Comment (alamb) [reply datafusion/physical-expr/src/expressions/in_list.rs]: Good idea -- done in 7cde104cf

**Comment by Dandandan** [comment]
Thanks @alamb !

**Pull request merged**